src/shred_collector/repair_service.zig

```diff
@@ -101,7 +98,7 @@ pub const RepairService = struct {
         var waiting_for_peers = false;
         var timer = try std.time.Timer.start();
         var last_iteration: u64 = 0;
-        while (!self.exit.load(.unordered)) {
+        while (!self.exit.load(.acquire)) {
```

Comment:
Why shouldn't the loop's contents be executed when `exit` is true? The function is run in a thread, which is joined before deinitialization happens. The join blocks the thread that's deiniting, waiting for the loop to see the `exit` flag and exit. 

It's necessary for the `exit` to be an `acquire` load in order to see modifications the parent thread does before releasing the exit flag being set.

Reply:
> It's necessary for the exit to be an acquire load in order to see modifications the parent thread does before releasing the exit flag being set.

I don't understand what you mean by "releasing the exit flag being set."

All that the "acquire" does is make sure that the atomic operation completes before the current thread executes the lines of code below the load.

Actually since this code branches based on the output of the load, I believe acquire semantics are already implied. I expect there is no difference between unordered and acquire in this case.

Reply:
> All that the "acquire" does is make sure that the atomic operation completes before the current thread executes the lines of code below the load.

This is, not what `acquire` means. A memory operation with `acquire` semantics creates a relationship with a pair `release`, on the same memory object. 

This means that if thread A does a release store to memory object M, and thread B sees that store using an acquire memory operation, in this specific case that would be the acquire load, it is guaranteed that all memory modifications done by thread A before that release store will be visible to thread B after the acquire load. 

This is critical to our exit flag, since it removes the possibility of some modification done by thread A being missed by thread B after it sees the exit flag "released", or set.


> Actually since this code branches based on the output of the load, I believe acquire semantics are already implied. I expect there is no difference between unordered and acquire in this case.

It is not implied anywhere, the path the code goes in has nothing to do with the memory operations. Atomics deals with the guaranteed ordering of visibility of memory between and inside of threads.

Please see my above godbolt link that shows exactly why an `unordered` load here would create an infinite loop, as the compiler will move the load outside of the while loop and it will never be updated.

Reply:
> This means that if thread A does a release store to memory object M, and thread B sees that store using an acquire memory operation, in this specific case that would be the acquire load, it is guaranteed that all memory modifications done by thread A before that release store will be visible to thread B after the acquire load.

This is consistent with my understanding. This is basically what I was getting at here:

> > All that the "acquire" does is make sure that the atomic operation completes before the current thread executes the lines of code below the load.
> 
> This is, not what acquire means.

It is what acquire means in the context of the thread executing the acquire operation, and it's compatible with your definition. I just didn't describe the full context of how it relates to a release, because I was only talking about the operations in the current thread that we're looking at now.

> This is critical to our exit flag, since it removes the possibility of some modification done by thread A being missed by thread B after it sees the exit flag "released", or set.

In the thread that sets exit to true, are there operations that need to happen before any other threads exit? I'm simply not aware of any scenarios where this is true, so I'm curious to hear about them.

In my previous comment, I was trying to imagine some scenarios where the ordering might matter. It was very generic/hypothetical, so I'm not sure if it applies, but the general idea was this:

Let's say thread A is going to trigger an exit with a store, and then thread B is going to load the exit variable and then exit after seeing it is true. If there are operations in thread A which need to complete before thread B exits, this can be handled by release/acquire semantics. But you might also want to ensure that some operations in thread B complete before thread A exits. This seems like it is potentially equally important, since exits can be triggered from anywhere, and it's not helped by the simple release/acquire ordering on exit store/loads.

> Please see my above godbolt link that shows exactly why an `unordered` load here would create an infinite loop, as the compiler will move the load outside of the while loop and it will never be updated.

This is a good argument against using unordered. I expect that monotonic would solve this specific problem. But actually it would be a more convincing argument if the code branched based on the output of the load, since that's how our code works. Something like this: https://godbolt.org/z/7e5vT3rW4. Since `cmp dword ptr [rdi], 0` is outside of `.LBB0_1:`, I think this confirms your argument about the infinite loop.

> > Actually since this code branches based on the output of the load, I believe acquire semantics are already implied. I expect there is no difference between unordered and acquire in this case.
>
> It is not implied anywhere, the path the code goes in has nothing to do with the memory operations. Atomics deals with the guaranteed ordering of visibility of memory between and inside of threads.

For some reason, I thought that the compiler would automatically use at least acquire ordering when the load is directly used in conditional branching logic, regardless of what the user specifies. But I think the godbolt example that I just shared proves me wrong. I looked around and could only find this on the topic: 

>When you use atomic variable directly in `if` or `switch` expression, implicit conversion happens, which is equivalent to `.load()` without parameters, and the memory order is implicitly `memory_order_seq_cst`. It is the safest memory order, so will not introduce any new issues.
>
>https://stackoverflow.com/questions/78298260/is-branching-on-a-stdatomic-variable-a-possible-source-of-bugs. 

Here they actually suggest sequential consistency is used, but again, this contradicts the example in godbolt, so I guess they are wrong.

Reply:
Sorry, limited to my phone for a couple of hours, so I’ll respond to the easier part. We can discuss the thread part later in our sync.

> For some reason, I thought that the compiler would automatically use at least acquire ordering when the load is directly used in conditional branching logic, regardless of what the user specifies.
…
Here they actually suggest sequential consistency is used, but again, this contradicts the example in godbolt, so I guess they are wrong. Maybe it's specific to C++.

This is specifically a C++ thing. `std::atomic` is a wrapper for a type, which like the more familiar `Arc<T>` in Rust, forwards use/deref to the inner type. In the C++ case, the use *must* load, so it does with `seq_cst` to ensure total order for edge case types.


Thankfully we don’t have terrible things like implicit total order in Zig.

Reply:
Commenting for posterity, to close the loop regarding my main concern:

> In the thread that sets exit to true, are there any specific modifications that actually need to happen before any of the other threads exit? I'm simply not aware of any scenarios where this is true, so I'm curious to hear about them.

No, there are no known scenarios where we actually need this. But it could conceivably be useful in the future, and there is no downside to including it at this point.